CI: GitHub Action migration from Jenkins CI

leo-automation opened the pull request:
This PR ports the existing Jenkins CI pipeline to GitHub Actions, aiming for functional parity while consolidating CI on GitHub.
### Approach
- **`prepare` job (build-only runner):** Pulls the Docker image from the internal registry and saves it as a chunked `.tar` artifact due to network restrictions on GPU runners. Will be moved to a GPU runner once we have some in internal network.
- **`build_and_test` job (GPU runner):** Downloads the artifact, loads the image into the node's Docker cache, and then starts a container. All subsequent build (`pip wheel` + `pip install`) and test steps run inside this container using `docker exec`, closely mirroring the single-agent Jenkins flow.
- Test execution uses the same scripts and structure (including parallel `sGPU` tests) as the original Jenkinsfile.

This migration focuses on directly replicating the existing CI logic in GitHub Actions.

leo-automation committed:
First drawt of the workflow

leo-automation committed:
Update transformer-engine-ci.yml

leo-automation committed:
Fixes

leo-automation committed:
Update transformer-engine-ci.yml

leo-automation committed:
Update transformer-engine-ci.yml

leo-automation committed:
toLower

leo-automation committed:
CI fixes

leo-automation committed:
Added build-only runners

leo-automation committed:
Debug docker credentials

leo-automation committed:
Debug docker credentials

leo-automation committed:
Node js is missing on the node

leo-automation committed:
Remove devcontainer

leo-automation committed:
More fixes

leo-automation committed:
Typo

leo-automation committed:
Permissions inside the container

leo-automation committed:
sudo

leo-automation committed:
Update transformer-engine-ci.yml

leo-automation committed:
Update transformer-engine-ci.yml

leo-automation committed:
Update transformer-engine-ci.yml

leo-automation committed:
Update transformer-engine-ci.yml

leo-automation committed:
Update transformer-engine-ci.yml

leo-automation committed:
Update transformer-engine-ci.yml

leo-automation committed:
Update transformer-engine-ci.yml

leo-automation committed:
Update transformer-engine-ci.yml

leo-automation committed:
Update transformer-engine-ci.yml

leo-automation committed:
Update transformer-engine-ci.yml

leo-automation committed:
Update transformer-engine-ci.yml

Commit:
CI: add CMAKE_CXX_COMPILER and CMAKE_CC_COMPILER

Commit:
DCMAKE_CXX_COMPILER

Commit:
ls -l

Commit:
view setup.py

Commit:
add dcmake cxx compiler to setup.py

Commit:
absolute path to dcmake_cxx_compiler

Commit:
hip_arch

Commit:
hip-runtime-amd and hip-dev

Commit:
dhip root dir

Commit:
rocm_path

Commit:
add HIP_PATH

Commit:
add setuptools and wheel

Commit:
tree

Commit:
tree

Commit:
rm hip dir

Commit:
back to basic

Commit:
rm build-essential

leo-automation committed:
Run CI

leo-automation committed:
Fix the path

leo-automation committed:
Test rootless

leo-automation committed:
Debug

leo-automation committed:
sudo

leo-automation committed:
sudo debug

leo-automation committed:
sudo

leo-automation committed:
AMDGPU_TARGETS

leo-automation committed:
Fix

leo-automation committed:
Debugiing

leo-automation committed:
Debugging

leo-automation committed:
Debugging

leo-automation committed:
PATH problem perhaps

leo-automation committed:
Debugging

leo-automation committed:
Debug

leo-automation committed:
Runner change

leo-automation committed:
Fix cplus

leo-automation committed:
Fix

leo-automation committed:
Fix

leo-automation committed:
Correct cmaker path

leo-automation committed:
Debug

leo-automation committed:
Cmake link

leo-automation committed:
Remove -f

leo-automation committed:
Debug

leo-automation committed:
PATH

leo-automation committed:
Refactoring to run container as a step

leo-automation committed:
build-only

leo-automation committed:
Houskeeping

leo-automation committed:
All in one stage

leo-automation committed:
GPU runner doesn't have access to internal network, so moved all network actions to build-only

leo-automation committed:
Split docker image

leo-automation committed:
Fix /opt/cmake

leo-automation committed:
Housekeeping

leo-automation committed:
Typo

leo-automation committed:
Pip install incorrect path

leo-automation committed:
Houskeeping

ipanfilo commented on .github/workflows/rocm-ci.yml:
it needs to be named something like rocm-ci.yml and have copyright header

leo-automation replied:
Done

ipanfilo replied:
Why does it have NVIDIA copyright?

leo-automation replied:
@ipanfilo, I took the same copyright as in the files in the same folder. Looked strange to me, but I figured out since it's NVIDIA's fork, it should be correct

ipanfilo replied:
Other files have NVIDIA copyright because their are created and maintained by NVIDIA. This file is for ROCm fork and owned by AMD so it should have corresponding copyright. 
For reference: https://github.com/ROCm/TransformerEngine/blob/dev/ci/_utils.sh

leo-automation replied:
Got it, thank you for clarifying. Updated the copyright

ipanfilo commented on .github/workflows/transformer-engine-ci.yml:
```diff
@@ -0,0 +1,229 @@
+          echo "Determining image for branch: $BRANCH_NAME"
+          DEV_DOCKER_IMAGE="compute-artifactory.amd.com:5000/rocm-plus-docker/framework/private/te-ci:rocm-6.4_0_ubuntu22_py310_torch25_jax0435qa_fa273"
+          REL613_DOCKER_IMAGE="compute-artifactory.amd.com:5000/rocm-plus-docker/framework/private/te-ci:rocm-6.4_0_ubuntu22_py310_torch25_jax0435qa_fa273"
+          LEGACY_DOCKER_IMAGE="compute-artifactory.amd.com:5000/rocm-plus-docker/framework/private/te-ci:rocm-6.3_66_ubuntu22_py310_torch25_jax0431_fa263_hipblaslt10patch.2"
```
Let's remove LEGACY images support. It needs to be also removed in Jenkins pipeline and stop monitoring corresponding branches

leo-automation replied:
Done. Both here and in the Jenkins CI

ipanfilo commented on .github/workflows/transformer-engine-ci.yml:
```diff
@@ -0,0 +1,229 @@
+          
+          export HIP_PATH=""
+          export PATH=/opt/cmake/bin:$PATH
+          export PYTORCH_ROCM_ARCH=gfx942
```
GPU arch should be variable or detected on runner. Otherwise it won;t work when switch to GFX950 runner for example

leo-automation replied:
Done with rocminfo command

ipanfilo commented on .github/workflows/transformer-engine-ci.yml:
```diff
@@ -0,0 +1,229 @@
+          #!/usr/bin/bash
+          set -x -o pipefail
+
+          HIP_VISIBLE_DEVICES=1 JUNITXML_PREFIX=junit_reports/torch_ JUNITXML_SUFFIX=.xml ci/pytorch.sh > /workspace/torch_sgpu.log 2>&1 &
```
I think JUNITXML is removed from pipeline. It is also moved to pipline level to run core,jax and torch stages in parallel

leo-automation replied:
Yep, removed those

Also, I don't think Github Actions allows to configure pipeline-level parallel run for steps, only for jobs. That wouldn't work in our case as each job needs its own runner

ipanfilo commented on .github/workflows/rocm-ci.yml:
```diff
@@ -0,0 +1,229 @@
+
+      - name: Cleanup container
+        if: always()
+        run: docker rm -f te-runner || true
```
should container rather be run with --rm?

leo-automation replied:
Usually, yes, but this workflow relies on starting the container and then running commands inside it with `docker exec`. Having an `always` step will remove the container even if the job failed

ipanfilo replied:
I think docker exec works only with started container. Original docker run command starts docker with default entry point (bash?) and it remains started till entry point app is completed or docker stop is called. In your case you just remove it w/o stopping. So, yes, if call with --rm, it will require explicit docker stop call

ipanfilo replied:
BTW, if run docker with something sh -c "sleep XXX" and run actual tests in docker exec, CI won't run longer than XX second, which helps to mitigate tests hangs.

leo-automation replied:
This is an interesting point, essentially, we'll create a timeout for a container. Though there's a GHA-native way using `timeout-minutes` with a step I am planning to implement, we can use it in the future when we need multiple timeouts inside one step

ipanfilo commented on .github/workflows/rocm-ci.yml:
```diff
@@ -0,0 +1,229 @@
+          EOF
+          )"
+
+      - name: Run sGPU tests
```
Before running tests I suggest to add ulimit -c 0 so gpucore dumps are not created - this problem is observed on Jenkins too

leo-automation replied:
Added

ipanfilo commented on .github/workflows/rocm-ci.yml:
```diff
@@ -0,0 +1,229 @@
+          name: docker-image-chunks
+
+      - name: Reassemble and Load Image to Node Cache
+        run: |
```
Will it it improve speed if images are moved to publicly accessible repository?

leo-automation replied:
@ipanfilo  Yes, for now. We are working on providing GPU runners with access to internal network, after that there should be no difference

leo-automation committed:
Resolving comments

leo-automation committed:
New line at the end

leo-automation committed:
Small changes

leo-automation committed:
Consolidate under one runner  with internal network access

ipanfilo committed:
Fix FFI import. Add distributed tests hang workaround (#347)

mkunredd committed:
Update rocm-ci.yml

add --network=host on docker run command

leo-automation committed:
Submodules

ipanfilo commented on .github/workflows/rocm-ci.yml:
```diff
@@ -0,0 +1,219 @@
+            -e GPU_ARCH=${{ steps.gpu-arch.outputs.arch }} \
+            te-runner bash -c "$(cat <<'EOF'
+          set -ex
+          apt-get update && apt-get install -y build-essential
```
CI images have build infrastructure installed so everything up to "pip wheel ."  seems not needed. Similar code is also removed from Jenkins

leo-automation replied:
Removed the code

ipanfilo committed:
Make TE ROCm wheels building image directly from manylinix image (#340)

* Build ROCm wheels directly from manylinix image
* Fix build on top of the latest Manylinix image
* Fix build after switching to AITER

leo-automation committed:
Added timeout

leo-automation committed:
Resolve comments

VeeraRajasekhar committed:
[CI] Hotfix test_gemm_autotune update (#353)

alextmagro committed:
MXFP8 test scale off by 1 fix (#338)

* MXFP8 test scale off by 1 fix

VeeraRajasekhar committed:
[CI] Removed Jax jit workaround, replaced with XLA_FLAGS=--xla_gpu_enable_nccl_comm_splitting=false (#346)

leo-automation committed:
Run CI

leo-automation committed:
Runner update

leo-automation committed:
Update runners

leo-automation committed:
Debug

leo-automation committed:
Debug

alextmagro committed:
Remove HIP macros around std:: math functions (#343)

* Remove HIP macros around std:: math functions
* hipify-torch commit

leo-automation committed:
Run CI

leo-automation committed:
Update runners label

VeeraRajasekhar committed:
[TE] [AITER] Add prebuilt AITER download and upload flow (#335)

* Implement prebuilt AITER binary download system via NVTE_AITER_PREBUILT_BASE_URL

* Add caching based on ROCm version and AITER commit SHA

* Implement SHA256 verification for downloaded prebuilt files

* Add automatic fallback to source build when prebuilt files unavailable

* Add cache validation to detect and clean outdated/invalid builds

* Implement upload file generation script for creating prebuilt files

* Addressed Reviews

* Fix git safe directory issues for AITER submodule in Docker environments

alextmagro committed:
std::max type mismatch hotfix (#361)

ipanfilo committed:
CI: allow numpy 2.0 (#366)

ipanfilo committed:
Relax tolerance to pass 29x29x17389NT GEMM on MI350 (#365)

Micky774 committed:
FIX Occasional import error when only building for a single framework (#355)

* Expanded catch categories

* Improved errors

* Converted assertion error to FileNotFound error for clarity

VeeraRajasekhar committed:
Integrate aiter HD192_HD128 backward kernels (#364)

Update 3rdparty/aiter commit which adds FMHAv3 backward support
for MLA head_dim_qk=192, head_dim_v=128 configuration.

- Add MLA test configs mla_4_0 and mla_4_1 for HD192_HD128
- Fix is_training logic to enable backward pass testing for HD192_HD128

* Added Jax fused_attn tests for MLA HD192+HD128 for bhsd

* Addressed reviews

ipanfilo committed:
Use .info/version for ROCm verison (#368)

alextmagro committed:
Enable aligned vectorized memory ops for MXFP8 cast (#342)

* Enable aligned vectorized memory ops for MXFP8 cast

* Optimized vector sizes and alignment conditions

wangye805 committed:
[ROCm] align the softmax aux shape with NVTE upstream (#371)

sudhu2k committed:
Te2.4 fsdp2 fp8 allgather autocast (#349)

* Initial commit

* Removed Print statements, added keep_fp8_transpose cache integration with fsdp2

* Added use_fsdp flag to Linear module, added profile code, added test code, added all reduce for amax

* Fixed unit test

* Removing all reduce code for amax since by default TE does all reduce when torch.distributed is initialized.

* reverting case where out is already present

* Added unit test with regualr sgpu training

* Modified unit test to compare FSDP2 with DDP

* bug fixes

* Code cleaning up

* Initial commit to add MXFP8

* Added fp8 current scaling.

* Added MXFP8, Modified unit test to run based on recipes

* Extended use_fsdp to layernorm linear and layernorm mlp

* Moved amax reduce from forward to backward for fsdp2

* Added automatic detection of use fsdp from base module

* Use SKIP_FP8_REDUCTION_FOR_FSDP2 in backward for check if need to do forward reduce

* Added memory profile code, added a check before setting SKIP_FP8_REDUCTION_FOR_FSDP2

* Fix for fused optimizer, changed _elem to _data, code clean up

* Fixed layernorm mlp

* Code cleanup and added test to pytorch.sh

* Removed whitespaces

* Fixed comments and license

* Added guards

* Added reduce for forward in cuda graph backward, added code to remove test artifacts, reverted upstream test file

---------

Co-authored-by: sudhu2k <sugovind@amd.com>

leo-automation committed:
Test CI plus small changes

leo-automation committed:
Run CI

leo-automation committed:
Update runners label

leo-automation committed:
Update runners

matthiasdiener committed:
[TE] Implement Triton current scaling (#341)

VeeraRajasekhar committed:
Update benchmark script to support fwd_v3 and a16 (#373)

* Restructured benchmark script, added new test

* Added tflops calculation

* Added casual factor in tflops calculation

* Added fwd_v3 argument to ci script

ipanfilo committed:
Enable AITER V3 kernels by default (#372)

leo-automation committed:
Add new logic from Jenkins and continue-on-error: true for tests

leo-automation committed:
Merge branch 'leo/migrate-ci-to-gha' of https://github.com/ROCm/TransformerEngine into leo/migrate-ci-to-gha
